### PR TITLE
scheduler: makes pod less racing so as to be put back into activeQ properly

### LINKED ISSUE #72925: Scheduler leaves pod stuck in pending after worker node reload
**What happened**:

Kubernetes scheduler may leave a single pod stuck in pending state after a worker node is reloaded/reinstalled. This problem appears to have been introduced in Kubernetes v1.12.4 and v1.13.1.  We have not been able to recreate the problem in our testing on earlier releases.  The key to the failure is that the worker being reloaded/reinstalled must be the only scheduling option available. The pending pod is eventually scheduled if a new scheduler leader is elected or the node status changes. This appears to be a timing bug since the problem does not always occur.

**What you expected to happen**:

No pods are stuck in pending state.

**How to reproduce it (as minimally and precisely as possible)**:

Start with a single node cluster running a handful of pods.  Cordon, drain and then reload/reinstall the node.  Since this is a timing bug, it may take several iterations to recreate the problem.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.4, 1.13.1 and 1.13.2
- Cloud provider or hardware configuration: VMs running in VirtualBox
- OS (e.g. from /etc/os-release): `Ubuntu 18.04.1 LTS`
- Kernel (e.g. `uname -a`): `Linux carrier0-master-1 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: Custom ansible scripts
- Others: N/A

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

`receivedMoveRequest` is a flag in internal scheduler queue structure set to `True` when the cluster becomes “more schedulable” upon node status changes. And the main benefit is to ensure current scheduling pod (which may be sentenced to be unscheduable as it used an earlier scheduler cache) has the right chance to be put back into activeQ (instead of unschedulableQ).

But in very rare cases the currently scheduling pod may lose the chance to be put back into activeQ. For example, Pod i is sentenced to be unschedulable, and then put into a goroutine ready to put back into scheduler queue (in factory.go#MakeDefaultErrorFunc()):

https://github.com/kubernetes/kubernetes/blob/4a42d97758918a47a08c6b5e19298b4d700a24c4/pkg/scheduler/factory/factory.go#L1114

Unfortunately it’s an asynchronous call, so it’s possible at that moment the cluster becomes “more schedulable” (e.g. some taints are removed), hence `receivedMoveRequest` is set to True and all pods in unschedulableQ are moved to activeQ – not including Pod i as it’s not in either queue. Next:
-	if the aforementioned Pod i’s goroutine runs, everything is fine as Pod i enjoys the benefit of  `receivedMoveRequest` and hence added to activeQ properly.
-	but it might be another case: the main scheduling goroutine runs first – which pops a pod from activeQ and immediately set `receivedMoveRequest` to False. This is not good, b/c Pod i loses the benefit of `receivedMoveRequest` and will be put to unschedulableQ.

I believe the issue has existed for several releases, but it’s not that severe until recently is b/c in the before, scheduler moves pods from unschedulabelQ to activeQ unconditionally, upon individual node status update – which occurs every 10 seconds. However recently we introduced PR #71551 to only move pods when the cluster becomes “more schedulable”. In other words, the issue is more critical b/c the aforementioned pod which was incorrectly put into unschedulableQ might be stuck for a long time, until the next “cluster more schedulable” event comes, and I believe this is what #72925 described.

**Which issue(s) this PR fixes**:

Fixes #72925 

**Special notes for your reviewer**:

As Bobby commented, this PR is unable to fix all racy conditions. For example, when Pod i gets wrapped into its goroutine, before it’s executed successfully, it’s possible (although pretty rare) Pod i+1 is sentenced to be unschedulable (as the event that cluster becomes “more schedulable” may be beneficial to Pod i, but not to Pod i+1) and be executed successfully. In this case, Pod i+1 sets flag `receivedMoveRequest` to False, and afterwards Pod i will be put into unschedulableQ. So PR #72558 which looks at unschedulableQ every 30 seconds is still needed.

**Does this PR introduce a user-facing change?**:

```release-note
scheduler: makes pod less racing so as to be put back into activeQ properly
```

/priority important-soon
/sig scheduling

### CODE REVIEW

**Review by bsalamat — COMMENTED** [review]
The title of your PR is confusing, but it makes sense and makes setting `receivedMoveRequest` less racy. I doubt if we can make it completely race free though. So, we have [developed](https://github.com/kubernetes/kubernetes/pull/72558) ways to tolerate occasional races.

**Comment by Huang-Wei** [comment]
@bsalamat yes, it can't fix all racing conditions. I put detailed analysis/explanation in the issue, and the title has been updated.

cc/ @k82cn @resouer @ravisantoshgudimetla @misterikkit for more comments.

/assign @bsalamat @k82cn

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -377,6 +377,7 @@ func (p *PriorityQueue) backoffPod(pod *v1.Pod) {
 // p.receivedMoveRequest is false or to backoff queue if p.receivedMoveRequest
 // is true but pod is subject to backoff. In other cases, it adds pod to active
 // queue.
+// It also clears receivedMoveRequest to mark the beginning of a new scheduling cycle.
```
Comment (bsalamat) [inline pkg/scheduler/internal/queue/scheduling_queue.go]: I think you should remove "to mark the beginning..." from this comment. This is not the beginning of a new scheduling cycle.

Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Done.

**Review by resouer — APPROVED** [review]
The idea looks good except one nits in comment :-)

**Comment by krmayankk** [comment]
Curious , how do we detect that only certain node status changes like taints/labels are what makes the pod more schedulable versus other changes might be superfluous and wont affect it

**Comment by Huang-Wei** [comment]
@krmayankk please check PR #71551 for more details.

(BTW: sig-node also implemented a feature (#69753) to only report status change when it's more reasonable.)

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @Huang-Wei

**Comment by cofyc** [comment]
hi, in debugging another issue https://github.com/kubernetes/kubernetes/issues/73216#issuecomment-457494598, I encountered same issue. I tried another way to fix this race completely, do you think is right way to do? https://github.com/kubernetes/kubernetes/pull/73309

**Comment by Huang-Wei** [comment]
@cofyc I've also thought about a similar solution, but using `timestamp` instead of `int64`. I will find time to leave comments on your PR.

**Pull request merged**